Allow NX support to be hidden from HVM guests as a domain creation
parameter.

diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index 7671a1f..6e486d8 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -2952,6 +2952,14 @@ int hvm_set_efer(uint64_t value)
         return X86EMUL_EXCEPTION;
     }
 
+    if ( (value & EFER_NX) &&
+         !(v->domain->arch.hvm_domain.params[HVM_PARAM_NX_ENABLED]) )
+    {
+        gdprintk(XENLOG_WARNING, "Trying to set NXE when it was disabled\n");
+        hvm_inject_page_fault(0, 0);
+        return X86EMUL_EXCEPTION;
+    }
+
     if ( ((value ^ v->arch.hvm_vcpu.guest_efer) & EFER_LME) &&
          hvm_paging_enabled(v) )
     {
@@ -4285,6 +4293,12 @@ void hvm_cpuid(unsigned int input, unsigned int *eax, unsigned int *ebx,
         /* Only provide PSE36 when guest runs in 32bit PAE or in long mode */
         if ( !(hvm_pae_enabled(v) || hvm_long_mode_enabled(v)) )
             *edx &= ~cpufeat_mask(X86_FEATURE_PSE36);
+        /* Fix up NX-disabling HVM param */
+#if CONFIG_PAGING_LEVELS >= 3
+        if ( !v->domain->arch.hvm_domain.params[HVM_PARAM_PAE_ENABLED] ||
+             !v->domain->arch.hvm_domain.params[HVM_PARAM_NX_ENABLED] )
+#endif
+            __clear_bit(X86_FEATURE_NX & 31, edx);
         /* Hide data breakpoint extensions if the hardware has no support. */
         if ( !boot_cpu_has(X86_FEATURE_DBEXT) )
             *ecx &= ~cpufeat_mask(X86_FEATURE_DBEXT);
@@ -5527,6 +5541,10 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
 
             switch ( a.index )
             {
+            case HVM_PARAM_NX_ENABLED:
+                if ( !cpu_has_nx )
+                    goto param_fail;
+                break;
             case HVM_PARAM_CALLBACK_IRQ:
                 hvm_set_callback_via(d, a.value);
                 hvm_latch_shinfo_size(d);
diff --git a/xen/include/public/hvm/params.h b/xen/include/public/hvm/params.h
index 403bfd0..5c2600d 100644
--- a/xen/include/public/hvm/params.h
+++ b/xen/include/public/hvm/params.h
@@ -56,6 +56,9 @@
 
 #if defined(__i386__) || defined(__x86_64__)
 
+/* Should the guest be allowed to use NX bits in page tables? */
+#define HVM_PARAM_NX_ENABLED   7
+
 /* The guest is using 32 bit PV drivers.  Apply suitable translations
    to shared info and hypercall arguments. */
 #define HVM_PARAM_32BIT        8
